operators/pkg/controller/elasticsearch/driver/deletion.go

```diff
@@ -0,0 +1,192 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/elastic/cloud-on-k8s/operators/pkg/apis/elasticsearch/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/reconciler"
+	esclient "github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/client"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/migration"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/mutation"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/observer"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/pod"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/reconcile"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/restart"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+type PodDeletionHandler struct {
+	es                 v1alpha1.Elasticsearch
+	performableChanges *mutation.PerformableChanges
+	results            *reconciler.Results
+	observedState      observer.State
+	reconcileState     *reconcile.State
+	esClient           esclient.Client
+	resourcesState     *reconcile.ResourcesState
+	defaultDriver      *defaultDriver
+}
+
+func (d *PodDeletionHandler) HandleDeletions() error {
+	if len(d.performableChanges.ToDelete) == 0 {
+		// nothing to do
+		return nil
+	}
+
+	// Start by deleting pods whose PVC should be kept around
+	if err := d.deleteForPVCReuse(d.performableChanges.ToDelete.WithPVCReuse()); err != nil {
+		return err
+	}
+
+	// Then delete pods which will not be replaced hence need data to be migrated
+	if err := d.deleteWithDataMigration(d.performableChanges.ToDelete.WithoutPVCReuse()); err != nil {
```

Comment:
> If you disable shard allocation for the reuse algo (even if its only replicas) how are you going to migrate data away from nodes that have data on them.

Yes, I also thought about that. We discussed it very briefly with @nkvoll, and realized maybe it's fine to have both in parallel: PVC reuse will temporarily block ongoing data migrations, but that's supposed to be temporary. But thinking this through again, you're right it does not work well. If we use the maxSurge budget to create new nodes (grow and shrink), effectively taking the place of a node to replace (pvc reuse), then we might have a deadlock.

> So while reuse is running (i.e. shard allocation disabled) no attempt to delete with migration and vice versa.

Would you check a "reuse is running" from retrieving ES shards allocation setting?
It's quite hard to know we are in a reuse phase in the current implementation, there is technically no difference between creating a pod for grow and shrink and creating a pod for PVC reuse (which, by design, sounds nice?).
I'm wondering if persisting a state in the cluster annotations would help here.

> Also sound like we are missing an e2e test for that

+1. I planned to add it with https://github.com/elastic/cloud-on-k8s/issues/1156 (`https://github.com/elastic/cloud-on-k8s/issues/1156`), but maybe it's major enough to deserve being in this PR.

Reply:
> I'm wondering if persisting a state in the cluster annotations would help here.

I was trying to find a way without resorting to annotations, but in the end it could be necessary to introduce one. (see my other comment)

> but maybe it's major enough to deserve being in this PR.

I think I am fine with doing it in #1156